dotnet/csharp-load/load/LoadGenerator.cs

```diff
@@ -0,0 +1,171 @@
+    private readonly string endpoint;
+
+    public LoadGenerator(CancellationTokenSource cancellationTokenSource, string endpoint, string authToken, string cache)
+    {
+        this.cancellationTokenSource = cancellationTokenSource;
+        this.endpoint = endpoint;
+        headers = new Grpc.Core.Metadata();
+        headers.Add(new Grpc.Core.Metadata.Entry("authorization", authToken));
+        headers.Add(new Grpc.Core.Metadata.Entry("cache", cache));
+    }
+
+    public async Task Run(uint hz, uint reportInterval, uint clients, uint concurrency) {
+        var cancellationToken = cancellationTokenSource.Token;
+
+        var metricsChannel = Channel.CreateUnbounded<Snapshot>();
+
+        var jobs = new List<Task>();
+        for (int i = 0; i < clients; ++i) {
+            jobs.Add(RunLoadAsync(i, hz / clients, concurrency, metricsChannel.Writer, cancellationToken));
+        }
+        jobs.Add(PrintPeriodicMetrics(cancellationToken, metricsChannel.Reader, TimeSpan.FromSeconds(reportInterval)));
+
+        await Task.WhenAny(jobs);
+        if (!cancellationToken.IsCancellationRequested) {
+            Console.WriteLine("Aborting load generation");
+            cancellationTokenSource.Cancel();
+        }
+
+        await Task.WhenAll(jobs);
+        Console.WriteLine("Load is all done");
+    }
+
+    private static async Task PrintPeriodicMetrics(CancellationToken cancellationToken, ChannelReader<Snapshot> reader, TimeSpan interval)
+    {
+        cancellationToken.Register(() => {
+            Console.WriteLine("Task metrics emitter: Cancelled");
+        });
+        Console.WriteLine("Starting metrics emitter");
+
+        var accumulation = Snapshot.Default();
+        
+        var lastReport = DateTime.Now;
+        while (!cancellationToken.IsCancellationRequested) {
+            var snapshot = await reader.ReadAsync(cancellationToken);
+            accumulation.Add(snapshot);
+
+            var now = DateTime.Now;
+            if (lastReport + interval < now) {
+                accumulation.Print(now - lastReport);
+                accumulation.Reset();
+
+                lastReport = now;
+            }
+        }
+    }
+
+    private async Task RunLoadAsync(int number, uint hz, uint maxLocalConcurrency, ChannelWriter<Snapshot> snapshotChannel, CancellationToken cancellationToken) {
+        cancellationToken.Register(() => {
+            Console.WriteLine("Task {0:D}: Cancelled", number);
+        });
+        Console.WriteLine("Starting {0:D} at {1:D}hz", number, hz);
+
+        var statSnapshot = Snapshot.Default();
+        const int operationCount = 2;
+
+        var lastReport = DateTime.Now;
+        var interval = new TimeSpan(TimeSpan.TicksPerSecond / hz) * operationCount;
+        var schedule = DateTime.Now;
+        var tasks = new HashSet<Task<Stats>>((int)maxLocalConcurrency);
+        ulong i = 0;
+
+        var channel = new Grpc.Core.Channel(endpoint, Grpc.Core.ChannelCredentials.SecureSsl);
```

Comment:
I thought you told me in slack that you needed to override some setting here in order to force it to actually use distinct channels/connections?

Reply:
Yeah, this change is stale. People did not review it for _checks notes_ 6 days so it has moved on. I'll update it soon.